planning/mission_planner/src/mission_planner/route_selector.cpp

```diff
@@ -0,0 +1,218 @@
+  mrm_operating_ = false;
+  main_request_ = std::monostate{};
+}
+
+void RouteSelector::on_state(const RouteState::ConstSharedPtr msg)
+{
+  if (msg->state == RouteState::UNSET && !initialized_) {
+    const auto stamp = now();
+    main_.change_state(stamp, RouteState::UNSET);
+    mrm_.change_state(stamp, RouteState::UNSET);
+    initialized_ = true;
+  }
+
+  (mrm_operating_ ? mrm_ : main_).update_state(*msg);
+}
+
+void RouteSelector::on_route(const LaneletRoute::ConstSharedPtr msg)
+{
+  (mrm_operating_ ? mrm_ : main_).update_route(*msg);
+}
+
+void RouteSelector::on_clear_route_main(
+  ClearRoute::Request::SharedPtr req, ClearRoute::Response::SharedPtr res)
+{
+  // Save the request to resume from MRM.
+  main_request_ = std::monostate{};
+
+  // During MRM, only change the state.
+  if (mrm_operating_) {
+    main_.change_state(now(), RouteState::UNSET);
+    res->status.success = true;
+    return;
+  }
+
+  // Forward the request if not in MRM.
+  res->status = service_utils::sync_call(cli_clear_route_, req);
+}
+
+void RouteSelector::on_set_waypoint_route_main(
+  SetWaypointRoute::Request::SharedPtr req, SetWaypointRoute::Response::SharedPtr res)
+{
+  // Save the request to resume from MRM.
+  req->uuid = uuid::generate_if_empty(req->uuid);
+  main_request_ = req;
+
+  // During MRM, only change the state.
+  if (mrm_operating_) {
+    main_.change_state(now(), RouteState::SET);
+    res->status.success = true;
+    return;
+  }
+
+  // Forward the request if not in MRM.
+  res->status = service_utils::sync_call(cli_set_waypoint_route_, req);
+}
+
+void RouteSelector::on_set_lanelet_route_main(
+  SetLaneletRoute::Request::SharedPtr req, SetLaneletRoute::Response::SharedPtr res)
+{
+  // Save the request to resume from MRM.
+  req->uuid = uuid::generate_if_empty(req->uuid);
+  main_request_ = req;
+
+  // During MRM, only change the state.
+  if (mrm_operating_) {
+    main_.change_state(now(), RouteState::SET);
+    res->status.success = true;
+    return;
+  }
+
+  // Forward the request if not in MRM.
+  res->status = service_utils::sync_call(cli_set_lanelet_route_, req);
+}
+
+void RouteSelector::on_clear_route_mrm(
+  ClearRoute::Request::SharedPtr req, ClearRoute::Response::SharedPtr res)
+{
+  mrm_operating_ = false;
+  mrm_.change_state(now(), RouteState::UNSET);
+
+  // Resume main route using the saved request.
+  if (std::holds_alternative<std::monostate>(main_request_)) {
+    res->status = service_utils::sync_call(cli_clear_route_, req);
+    return;
+  }
+  if (auto request = std::get_if<RoutePointRequest>(&main_request_)) {
+    // NOTE: Clear the waypoint to avoid returning. Remove this once resuming is supported.
+    (**request).waypoints.clear();
+    res->status = service_utils::sync_call(cli_set_waypoint_route_, *request);
+    return;
+  }
+  if (auto request = std::get_if<RouteRequest>(&main_request_)) {
+    res->status = service_utils::sync_call(cli_set_lanelet_route_, *request);
+    return;
+  }
```

Comment:
In the original code, even in the case of RouteRequest, if the reroute fails, it is retried using only the goal.

Reply:
Thank you. I fixed it to call `set_waypoint_route` with only goal in both cases. https://github.com/autowarefoundation/autoware.universe/pull/6363/commits/09a753374ce85dc786a242fccd385eff0df88bb3

Reply:
thanks and sorry.

In the original code, 
1. Check if vehicle can return to the original normal root  https://github.com/autowarefoundation/autoware.universe/blob/main/planning/mission_planner/src/mission_planner/mission_planner.cpp#L510-L516
2. If not, check with a new route using only goal. https://github.com/autowarefoundation/autoware.universe/blob/main/planning/mission_planner/src/mission_planner/mission_planner.cpp#L518-L530

Reply:
same for WaypointRequest,
1. check original normal route
2. check new route with only goal

But the route_selector does not keep the original normal route, so it may be a bit difficult.

Reply:
Fixed so that it first attempts with the planned lanelet segments and then replan with the goal. https://github.com/autowarefoundation/autoware.universe/pull/6363/commits/9fe0af236f554ee9c47b39cbd3af8443854248c3